nemo/collections/asr/modules/rnn_encoder.py

```diff
@@ -0,0 +1,187 @@
+
+    @property
+    def input_types(self):
+        """Returns definitions of module input ports.
+        """
+        return OrderedDict(
+            {
+                "audio_signal": NeuralType(('B', 'D', 'T'), SpectrogramType()),
+                "length": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    @property
+    def output_types(self):
+        """Returns definitions of module output ports.
+        """
+        return OrderedDict(
+            {
+                "outputs": NeuralType(('B', 'D', 'T'), AcousticEncodedRepresentation()),
+                "encoded_lengths": NeuralType(tuple('B'), LengthsType()),
+            }
+        )
+
+    def __init__(
+        self,
+        feat_in,
+        n_layers,
+        d_model,
+        proj_size=-1,
+        rnn_type='lstm',
+        bidirectional=True,
+        subsampling='striding',
+        subsampling_factor=4,
+        subsampling_conv_channels=-1,
+        dropout=0.1,
+    ):
+        super().__init__()
+
+        self.d_model = d_model
+        self._feat_in = feat_in
+
+        if subsampling_conv_channels == -1:
+            subsampling_conv_channels = proj_size
+        if subsampling and subsampling_factor > 1:
+            if subsampling == 'stacking':
+                self.pre_encode = StackingSubsampling(
+                    subsampling_factor=subsampling_factor, feat_in=feat_in, feat_out=proj_size
+                )
+            else:
+                self.pre_encode = ConvSubsampling(
+                    subsampling=subsampling,
+                    subsampling_factor=subsampling_factor,
+                    feat_in=feat_in,
+                    feat_out=proj_size,
+                    conv_channels=subsampling_conv_channels,
+                    activation=nn.ReLU(),
+                )
+        else:
+            self.pre_encode = nn.Linear(feat_in, proj_size)
+
+        self._feat_out = proj_size
+
+        self.layers = nn.ModuleList()
+
+        SUPPORTED_RNN = {"lstm": nn.LSTM, "gru": nn.GRU, "rnn": nn.RNN}
+        if rnn_type not in SUPPORTED_RNN:
+            raise ValueError(f"rnn_type can be one from the following:{SUPPORTED_RNN.keys()}")
+        else:
+            rnn_module = SUPPORTED_RNN[rnn_type]
+
+        for i in range(n_layers):
+            rnn_proj_size = proj_size // 2 if bidirectional else proj_size
+            if rnn_type == "lstm":
+                layer = rnn_module(
+                    input_size=self._feat_out,
+                    hidden_size=d_model,
+                    num_layers=1,
+                    batch_first=True,
+                    bidirectional=bidirectional,
+                    proj_size=rnn_proj_size,
+                )
+            self.layers.append(layer)
+            self.layers.append(nn.LayerNorm(proj_size))
+            self.layers.append(nn.Dropout(p=dropout))
+            self._feat_out = proj_size
+
+        if proj_size > 0 and self._feat_out != proj_size:
```

Comment:
Ah ok final projection only, hmm then just overload the `pred_hidden` of `model_defaults` here.